[써니] 코틀린 레이싱카 제출합니다.

inseonyun opened the pull request:
안녕하세요. 이번 레이싱카 미션 진행하며, 궁금한 내용 남깁니다.

1. 페어와 함께 최대한 MVC 패턴을 적용하며 소스 작성을 하고자 노력하였는데, 전체적인 게임 진행을 담당하는 Controller와 세부 기능에 대한 결과를 도출하는 Service 그리고 View, Model을 구분하여 작성하였는데 부족한 점이나, 위배되는점이 있다면 코멘트 부탁드립니다...
2. 제 경우 코틀린스로운 코드를 작성하는게 가장 어려웠던 것 같습니다. 간결하지 못하고, for문을 사용해 쉼표를 찍어주거나, 우승자를 찾는다거나 하는 로직을 구성했었는데, 페어를 통해 이러한 점을 보완해 나갔었는데 그럼에도 불구하고 완성된 코드를 보면 코틀린스럽지 않아 보이기도 하고 많이 부족해 보입니다. 혹시 이건 굳이 이렇게 안 해도 될 거 같은데..? 라던가 좀 더 간결하게 짤 수 있는 방법이 있다면 코멘트 부탁드립니다.

감사합니다 :)

inseonyun committed:
docs: 요구사항 문서 작성

inseonyun committed:
feat: 자동차 이름 입력 예외 처리 클래스 작성

inseonyun committed:
feat: 시도 횟수가 숫자인지 check 하는 함수 구현

inseonyun committed:
feat: 시도 횟수가 양의 정수인지 check 하는 함수 구현

inseonyun committed:
refactor: 이름 입력 시 발생하는 예외 처리 클래스 수정

inseonyun committed:
refactor: 이름 입력 시 Null 발생 예외 처리 수정

inseonyun committed:
docs: 요구사항 수정

inseonyun committed:
feat: 이름 입력 기능 구현

inseonyun committed:
refactor: 입력 시 발생할 수 있는 Null 예외 처리 클래스 구현

inseonyun committed:
feat: 시도 횟수 입력 기능 구현

inseonyun committed:
feat: 안내 문구 출력 기능 구현

inseonyun committed:
feat: 라운드 결과 출력 기능 구현

inseonyun committed:
feat: 최종 우승자 출력 기능 구현

inseonyun committed:
feat: Car 클래스 생성

inseonyun committed:
feat: 무작위 수를 뽑는 함수 구현

inseonyun committed:
feat: 무작위 수가 4이상인지 확인하는 기능 구현

inseonyun committed:
feat: 자동차 이름 쉼표를 기준으로 스플릿하는 함수 구현

inseonyun committed:
feat: 라운드 별 결과를 확인하는 기능 구현

inseonyun committed:
feat: 우승한 자동차 이름을 반환하는 기능 구현

inseonyun committed:
feat: 레이싱 결과를 출력 형식으로 변환하는 기능 구현

inseonyun committed:
refactor: 함수명 변경

inseonyun committed:
feat: 컨트롤러 기능 구현

inseonyun committed:
refactor: 이름 입력 예외 처리 수정

inseonyun committed:
refactor: 입력 예외 발생 시 재입력 기능 구현

inseonyun committed:
refactor: 시도 횟수 예외 처리 함수 수정

inseonyun committed:
refactor: 이름 입력 예외 처리 수정

inseonyun committed:
refactor: 전체 코드 객체지향적으로 수정

inseonyun committed:
refactor: 테스트에 용이하게 코드 수정

inseonyun committed:
test: InputValidationTest 구현

inseonyun committed:
refactor: RandomGenerator를 함수 안에서 구현하지 않도록 수정

inseonyun committed:
feat: DomainTest 구현

BeokBeok commented on docs/README.md:
```diff
@@ -0,0 +1,58 @@
+
+## 입력
+- [x] 이름을 입력 받는다.
+- [ ] 시도 횟수를 입력 받는다.
```
```suggestion
- [x] 시도 횟수를 입력 받는다.
```
시도 횟수를 입력받고 있으니, 체크해보면 어떨까요?

BeokBeok commented on src/main/kotlin/controller/RacingGameController.kt:
```diff
@@ -0,0 +1,58 @@
+import view.OutputView
+
+class RacingGameController {
+    private val inputView = InputView()
+    private val outputView = OutputView()
+
+    private val racingCarGameService = RacingCarGameService()
+    private var carsInfo = listOf<Car>()
+    private var tryCount = 0
```
멤버 변수를 활용하여 코드를 구현해주셨네요.
멤버 변수를 활용하면 클래스 내부에서 값을 편하게 가져다 쓸 수 있는 장점이 있지만, 멤버 변수의 값에 따라 함수 호출 시 결과가 매번 달라지게 되서, 함수의 결괏값을 예측하기가 어려워져요.

멤버 변수에 의존하지 않고 동작할 수 있도록 코드를 개선해보면 어떨까요? (이하 관련 내용 동일)

ref. https://dev.to/adammc331/the-imposters-guide-to-dependency-injection-1eak

BeokBeok commented on src/main/kotlin/controller/RacingGameController.kt:
```diff
@@ -0,0 +1,58 @@
+        gerWinner()
+    }
+
+    fun getCarNames() {
+        try {
+            outputView.printCar()
+            val carsName = racingCarGameService.splitCarNames(inputView.inputName())
+            carsInfo = racingCarGameService.initCarInfo(carsName)
+        } catch (e: IllegalArgumentException) {
+            e.printStackTrace()
+            getCarNames()
+        }
+    }
```
함수명만 보았을 때에는 "자동차의 이름을 얻다"의 의미로 이해가 되는데요.
그래서 이 함수를 호출하게 되면 자동차의 이름이 리턴될 것으로 예상되지만, 실제로는 리턴값이 없네요.
함수의 의도를 좀 더 명확하게 작성해보면 어떨까요? (이하 관련 내용 동일)

BeokBeok commented on src/main/kotlin/controller/RacingGameController.kt:
```diff
@@ -0,0 +1,58 @@
+
+    fun playWholeRacing() {
+        outputView.printRunResultMessage()
+        for (round in 1..tryCount) playRound()
```
```suggestion
        repeat(tryCount - 1) {
            playRound()
        }
```
repeat 함수를 활용해보면 어떨까요?

BeokBeok commented on src/main/kotlin/controller/RacingGameController.kt:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    fun gerWinner() {
+        outputView.printWinner(racingCarGameService.getWinners(racingCarGameService.getMaxPositionCars(carsInfo)))
```
printWinner함수를 호출하는데, getWinners 함수를 호출하고 그 안에 getMaxPositionCars 함수를 호출하고 있네요.
함수를 인라인해서 호출을 잘 해주셨지만, 이와 같이 인라인을 하게 되면 각 함수의 호출 결과가 의미하는 바를 이해하기가 어렵게 됩니다.
이런 경우에는 변수를 활용해서 각 함수들이 호출하는 값이 의미하는 것을 드러내보면 어떨까요?

BeokBeok commented on src/main/kotlin/exception/NameException.kt:
```diff
@@ -0,0 +1,28 @@
+    private val nullException = NullException()
+    fun checkNames(carNames: String) {
+        splitCarNames(carNames).forEach { name ->
+            checkNameRange(nullException.checkNull(name))
```
```suggestion
            checkNameRange(name)
```
checkNames 함수의 파라미터(name)의 타입은 non-null 타입입니다.
즉, name은 null일 수 없다는 의미입니다.

BeokBeok commented on src/main/kotlin/exception/NameException.kt:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    fun checkNameRange(carName: String) {
+        if (carName.isEmpty() || carName.length > 5) {
```
숫자 5는 어떤 것을 의미할까요?
숫자 5와 같은 것은 "매직리터럴"이라고 합니다.
"매직리터럴"에 대해서 학습해보시고, 어떻게 개선하면 좋을지 고민해보면 어떨까요? (이하 관련 내용 동일)

BeokBeok commented on src/main/kotlin/exception/TryCountException.kt:
```diff
@@ -0,0 +1,23 @@
+
+import message.ExceptionMessage
+
+class TryCountException {
```
클래스명을 보니 Exception을 custom한 것으로 보이지만, 실제로는 Exception 타입이 아니네요.
이 클래스가 하는 책임에 대해서 다시 한번 고민해보면 어떨까요?
클래스명이 Exception이라서 마치 IllegalArgumentException과 같은 Exception의 한 종류로 보이지만, 실제로는 단순히 유효성을 체크하는 책임을 가지고 있는 것으로 보여요.

inseonyun replied:
해당 사항 Exception이 아닌 Validation으로 수정하였습니다.
예외 처리 관련해서 궁금한 것이 있습니다.
코틀린에선 requre와 check 를 통해 각각 IllegalArgumentException / IllegalStateException 예외를 발생 시킬 수 있다고 알고 있습니다.
제가 작성한 코드에서는 모두 IllegalArgumentException를 발생하고 있어, 조건문을 통해 throw를 하는 것이 아닌 require를 통해 예외를 처리하는 방법도 있을 것 같은데 조건문을 통한 throw와 require 둘 중 어떤 것이 좀 더 코틀린스러운? 선택인가요?

BeokBeok replied:
require 혹은 check를 사용하는 상황과, 조건문을 통해 throw하는 상황이 서로 다를 수는 있습니다.
가급적이면 require 혹은 check와 같은 선언적 체크문을 활용하는 것이, 가독성 측면이나 스마트케스팅 등 여러 이점이 있습니다.

BeokBeok commented on src/main/kotlin/message/ExceptionMessage.kt:
```diff
@@ -0,0 +1,11 @@
+package message
+
+class ExceptionMessage {
+    companion object {
```
```suggestion
object ExceptionMessage {
```
클래스 내부에 companion object만 활용할 것이라면, class 대신 object를 활용할 수 있어요 (이하 관련 내용 동일)

inseonyun replied:
동반 객체와 관련해 궁금한 것이 있습니다.
제 경우 이번 자동차 경주 미션 때와 마찬가지로 message를 갖는 상수들을 종류에 맞게 class로 묶어 하나의 class 에서 관리를 하는 방법으로 사용해 왔는데.
해당 message를 사용하는 클래스 내에서 동반 객체를 생성하여 상수를 사용하는 방법과 위 방법 중 어느 것이 관리, 코드 작성 하는데에 이점이 있을까요?? 리뷰어님과 실제 현업에서는 주로 어떤 방법으로 상수를 갖는 동반 객체를 관리하나요?

BeokBeok replied:
대체로 클래스는 최소한 하나 이상을 캡슐화를 해야합니다.
그렇지 않으면, 클래스를 사용하는 것 보다는 object를 사용하는 것을 권장합니다.
특히 상수값같은 경우에는, 여러 클래스에서 접근하는 경우에는 별도의 object를 만들어서 관리하고 있으며, 클래스 내에서만 사용하는 상수값이라면 companion object에 관리하고 있습니다.

BeokBeok commented on src/test/kotlin/DomainTest.kt:
```diff
@@ -0,0 +1,42 @@
+import org.junit.jupiter.params.provider.CsvSource
+import service.RacingCarGameService
+
+class DomainTest {
```
DomainTest는 어떤 것을 의미할까요?
Test 클래스는 가급적이면 테스트하고자 하는 클래스와 일대일 매칭시키는 것이 좋습니다.
가령, Car 클래스를 테스트하는 클래스를 만들고자 한다면, CarTest 가 된다고 보시면 됩니다.

BeokBeok reviewed (changes requested):
자동차 경주 구현 미션하시느라 고생하셨습니다. 👍 
고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪

BeokBeok commented:
> 페어와 함께 최대한 MVC 패턴을 적용하며 소스 작성을 하고자 노력하였는데, 전체적인 게임 진행을 담당하는 Controller와 세부 기능에 대한 결과를 도출하는 Service 그리고 View, Model을 구분하여 작성하였는데 부족한 점이나, 위배되는점이 있다면 코멘트 부탁드립니다...

A. 전반적으로 고민해볼만한 부분들을 코맨트로 작성하였습니다. 😅 

> 제 경우 코틀린스로운 코드를 작성하는게 가장 어려웠던 것 같습니다. 간결하지 못하고, for문을 사용해 쉼표를 찍어주거나, 우승자를 찾는다거나 하는 로직을 구성했었는데, 페어를 통해 이러한 점을 보완해 나갔었는데 그럼에도 불구하고 완성된 코드를 보면 코틀린스럽지 않아 보이기도 하고 많이 부족해 보입니다. 혹시 이건 굳이 이렇게 안 해도 될 거 같은데..? 라던가 좀 더 간결하게 짤 수 있는 방법이 있다면 코멘트 부탁드립니다.

A. 코틀린스러운 코드를 작성하는 것은 아직 익숙지 않아서 어려우신 것 같습니다.
코틀린과 좀 친해지다보면 자연스럽게 코틀린스러운 코드를 작성하실 수 있으실거에요. 💪

inseonyun committed:
refactor: 멤버 변수 초기화 의존성 수정

inseonyun committed:
refactor: 함수 네이밍 수정

inseonyun committed:
refactor: 매직 리터럴 수정

inseonyun committed:
refactor: 컨트롤러 수정

1. for문 -> repeat
2. outputWinner() 각 함수 결과 변수에 담은 후 출력

inseonyun committed:
refactor: 입력 예외 처리 클래스, 패키지명 수정

inseonyun committed:
refactor: 동반 객체만 갖는 class object로 수정

기존 validationMessage 각 validation 클래스에서 동반 객체를 갖도록 함

inseonyun committed:
refactor: Car 클래스 수정

생성자 name의 접근 제한 해제

inseonyun committed:
refactor: DomainTest 세분화

1. CarTest - 자동차 전진 및 현재 상태 테스트
2. RacingCarGameTest - 가장 멀리간 자동차 및 우승한 자동차 테스트

inseonyun committed:
refactor: NameValidation 수정

1. NullValidation 생성자 제거
2. checkNames()의 파라미터 name은 non-null이므로, name은 null일 수 없음

inseonyun committed:
refactor: Validation 및 View object로 수정

inseonyun committed:
refactor: for문 repeat 함수로 수정

BeokBeok commented on src/main/kotlin/controller/RacingGameController.kt:
```diff
@@ -0,0 +1,64 @@
+        outputWinner()
+    }
+
+    fun inputCarNames() {
+        try {
+            OutputView.printCar()
+            val carsName = racingCarGameService.splitCarNames(InputView.inputName())
+            carsInfo = racingCarGameService.initCarsInfo(carsName)
+        } catch (e: IllegalArgumentException) {
+            e.printStackTrace()
+            inputCarNames()
+        }
+    }
```
exception이 발생하면 해당 함수를 재귀로 호출하고 있네요.
지금과 같은 구조는 의도하신대로 계속해서 입력받을 수 있는 장점이 있지만, 프로그램이 예상치 못한 오류가 발생했을 때 어떤 오류가 발생했는지 모른 채로 프로그램이 종료되지 않고 실행되게 됩니다.
또한, 입력에 관한 책임은 InputView에 있으니, Controller가 아닌 InputView에서 처리해보면 어떨까요? (이하 관련 내용 동일)

BeokBeok commented on src/main/kotlin/controller/RacingGameController.kt:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    companion object {
+        const val DEFAULT_TRY_COUNT = 0
```
```suggestion
        private const val DEFAULT_TRY_COUNT = 0
```
클래스 내부에서만 접근하는 것이라면, private 접근자를 붙여보면 어떨까요? (이하 관련 내용 동일)

BeokBeok commented on src/main/kotlin/service/RacingCarGameService.kt:
```diff
@@ -0,0 +1,53 @@
+        return carsInfo.filter { maxCarInfo.isSamePosition(it) }
+    }
+
+    fun getRoundResult(carsInfo: List<Car>): String {
+        var roundResult = ""
+        carsInfo.forEach { car ->
+            roundResult += "${car.getPresentStatus()}\n"
+        }
+        return roundResult
+    }
+
+    fun getWinners(carsInfo: List<Car>): String {
+        var winners = "최종 우승자: "
+
+        carsInfo.forEach { car ->
+            winners += "${car.name}, "
+        }
+        return winners.substring(SUBSTRING_START_POSITION, winners.length - SUBSTRING_DELETE_COMMA_AND_SPACE)
+    }
```
출력과 관련된 문구를 만들어주고 있네요.
출력과 관련된 책임은 OutputView에서 가지고 있으니, OutputView에서 처리해보면 어떨까요?

BeokBeok commented on src/test/kotlin/RacingCarGameTest.kt:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @Test
+    fun `우승한 자동차를 정상적으로 출력하는지 테스트`() {
```
여기에서 "정상"이라는 의미는 어떤 의미일까요?

BeokBeok commented on src/test/kotlin/RacingCarGameTest.kt:
```diff
@@ -0,0 +1,31 @@
+    private val racingCarGameService = RacingCarGameService()
+
+    @Test
+    fun `가장 멀리간 자동차를 확인하는 테스트`() {
```
클래스명만 봐도 테스트를 의미하므로, 굳이 함수명에 "테스트"를 언급하지 않으셔도 됩니다. 😅

BeokBeok reviewed (approved):
피드백 반영하시느라 고생하셨습니다. 👍 
추가로 고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪 
다음 미션을 진행하시면서, 추가로 작성된 코맨트들도 반영해주세요. 🙏
